llvm/lib/Transforms/Scalar/LoopInterchange.cpp

```diff
@@ -204,11 +204,12 @@ static bool isLegalToInterChangeLoops(CharMatrix &DepMatrix,
   std::vector<char> Cur;
   // For each row check if it is valid to interchange.
   for (unsigned Row = 0; Row < NumRows; ++Row) {
-    // Create temporary DepVector check its lexicographical order
-    // before and after swapping OuterLoop vs InnerLoop
+    // A dependence vector has to be lexigraphically positive. We could assert
+    // that here, as a sanity check for the dependency analysis, but it can
+    // contain "don't know" elements ('*'), which will be rightfully
+    // interpreted as not lexicographical positive. So, skip that sanity check,
+    // it suffices just to check the interchanged direction vector.
```

Comment:
AFAIK, there was a reason for `isLexicographicallyPositive(Cur)` to be called before the swap. I did not recall explicitly but it likely has something to do with `'S'` or `'*'` dependency. We had encountered a number of bugs due to the fact that `'S'` and `'*'` not very easily dealt with in loop interchange. 

If you remove the `isLexicographicallyPositive()` call here, likely bugs would appear. I probably do not have enough time digging into it at the moment, but you may refer to this patch (`https://reviews.llvm.org/D137461`) and also earlier patches and issues in loop interchange.

Reply:
Thanks for that pointer. 

I agree with the following from that phabricator review, that's what I am saying too:

> At the LoopWG call we discussed that we still could call isLexicographicallyPositive before the interchange to check that the dependency is understood before risting breaking things. I don't think it's necessary, but I don't mind.

The impression that I get from that discussion and the mishandling of scalar dependencies, is that we are not really sure why we are doing things.

Anyway, I don't have too strong opinions on this, I still don't think it is necessary, but I will focus on #119345 first.